components/function-controller/pkg/apis/serverless/v1alpha1/function_validation.go

```diff
@@ -111,117 +127,119 @@ func (spec *FunctionSpec) validateEnv(ctx context.Context) (apisError *apis.Fiel
+		allErrs = append(allErrs, fmt.Sprintf("%s.limits.cpu(%s) should be higher than %s.requests.cpu(%s)",
+			parent, limits.Cpu().String(), parent, requests.Cpu().String()))
 	}
 	if requests.Memory().Cmp(*limits.Memory()) == 1 {
-		apisError = apisError.Also(apis.ErrInvalidValue(
-			fmt.Sprintf("limits memory(%s) should be higher than requests memory(%s)",
-				limits.Memory().String(), requests.Memory().String()), "limits.memory"))
+		allErrs = append(allErrs, fmt.Sprintf("%s.limits.memory(%s) should be higher than %s.requests.memory(%s)",
+			parent, limits.Memory().String(), parent, requests.Memory().String()))
 	}
-
-	return apisError
+	return returnAllErrs("invalid function resources", allErrs)
 }
 
-func (spec *FunctionSpec) validateReplicas(ctx context.Context) (apisError *apis.FieldError) {
-	minValue := ctx.Value(ValidationConfigKey).(ValidationConfig).Function.Replicas.MinValue
+func (spec *FunctionSpec) validateReplicas(vc *ValidationConfig) error {
+	minValue := vc.Function.Replicas.MinValue
 	maxReplicas := spec.MaxReplicas
 	minReplicas := spec.MinReplicas
-
+	allErrs := []string{}
 	if maxReplicas != nil && minReplicas != nil && *minReplicas > *maxReplicas {
-		apisError = apisError.Also(apis.ErrInvalidValue(
-			fmt.Sprintf("maxReplicas(%d) is less than minReplicas(%d)", *maxReplicas, *minReplicas), "spec.maxReplicas"))
+		allErrs = append(allErrs, fmt.Sprintf("spec.maxReplicas(%d) is less than spec.minReplicas(%d)",
+			*maxReplicas, *minReplicas))
 	}
 	if minReplicas != nil && *minReplicas < minValue {
-		apisError = apisError.Also(apis.ErrInvalidValue(
-			fmt.Sprintf("minReplicas(%d) is less than the smallest allowed value(%d)", *minReplicas, minValue), "spec.minReplicas"))
+		allErrs = append(allErrs, fmt.Sprintf("spec.minReplicas(%d) is less than the smallest allowed value(%d)",
+			*minReplicas, minValue))
 	}
 	if maxReplicas != nil && *maxReplicas < minValue {
-		apisError = apisError.Also(apis.ErrInvalidValue(
-			fmt.Sprintf("maxReplicas(%d) is less than the smallest allowed value(%d)", *maxReplicas, minValue), "spec.maxReplicas"))
+		allErrs = append(allErrs, fmt.Sprintf("spec.maxReplicas(%d) is less than the smallest allowed value(%d)",
+			*maxReplicas, minValue))
 	}
-
-	return apisError
+	return returnAllErrs("invalid values", allErrs)
 }
 
-func (spec *FunctionSpec) validateLabels() (apisError *apis.FieldError) {
+func (spec *FunctionSpec) validateLabels(_ *ValidationConfig) error {
 	labels := spec.Labels
 	fieldPath := field.NewPath("spec.labels")
 
 	errors := v1validation.ValidateLabels(labels, fieldPath)
-	for _, err := range errors {
-		apisError = apisError.Also(apis.ErrInvalidValue(err.Error(), "spec.labels"))
-	}
-	return apisError
+	return errors.ToAggregate()
 }
 
 type property struct {
 	name  string
 	value string
 }
 
-func validateIfMissingFields(properties ...property) (apisError *apis.FieldError) {
+func validateIfMissingFields(properties ...property) error {
```

Comment:
We can move this function to `validation_helpers.go` and other methods that are generic.